src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java

```diff
@@ -0,0 +1,440 @@
+  private int timeoutCheckPeriod = 1000;
+
+  private long flushHardTimeoutTimestamp;
+  private long flushSoftTimeoutTimestamp;
+  // The maximum soft timeout allowed. This is used while the app is shutting
+  // down to increase the likelihood of flushing before the app finishes
+  // shutting down.
+  private long flushMaximumSoftTimeout;
+
+  private final BackgroundFlushThread backgroundFlushThread = new BackgroundFlushThread();
+  private final BackgroundSyncThread backgroundSyncThread = new BackgroundSyncThread();
+
+  private boolean activated = false;
+
+  public AbstractBufferedRollingFileAppender () {
+    this(new SystemTimeSource());
+  }
+
+  /**
+   * Constructor
+   * @param timeSource The time source that the appender should use
+   */
+  public AbstractBufferedRollingFileAppender (TimeSource timeSource) {
+    this.timeSource = timeSource;
+
+    // The default flush timeout values below are optimized for high latency
+    // remote persistent storage such as object stores or HDFS
+    flushHardTimeoutPerLevel.put(Level.FATAL, 5L * 60 * 1000 /* 5 min */);
+    flushHardTimeoutPerLevel.put(Level.ERROR, 5L * 60 * 1000 /* 5 min */);
+    flushHardTimeoutPerLevel.put(Level.WARN, 10L * 60 * 1000 /* 10 min */);
+    flushHardTimeoutPerLevel.put(Level.INFO, 30L * 60 * 1000 /* 30 min */);
+    flushHardTimeoutPerLevel.put(Level.DEBUG, 30L * 60 * 1000 /* 30 min */);
+    flushHardTimeoutPerLevel.put(Level.TRACE, 30L * 60 * 1000 /* 30 min */);
+
+    flushSoftTimeoutPerLevel.put(Level.FATAL, 5L * 1000 /* 5 sec */);
+    flushSoftTimeoutPerLevel.put(Level.ERROR, 10L * 1000 /* 10 sec */);
+    flushSoftTimeoutPerLevel.put(Level.WARN, 15L * 1000 /* 15 sec */);
+    flushSoftTimeoutPerLevel.put(Level.INFO, 3L * 60 * 1000 /* 3 min */);
+    flushSoftTimeoutPerLevel.put(Level.DEBUG, 3L * 60 * 1000 /* 3 min */);
+    flushSoftTimeoutPerLevel.put(Level.TRACE, 3L * 60 * 1000 /* 3 min */);
+  }
+
+  /**
+   * Sets whether to close the log file upon receiving a shutdown signal before
+   * the JVM exits. If set to false, the appender will continue appending logs
+   * even while the JVM is shutting down and the appender will do its best to
+   * sync those logs before the JVM shuts down. This presents a tradeoff
+   * between capturing more log events and potential data loss if the log events
+   * cannot be flushed and synced before the JVM shuts down.
+   * @param closeFileOnShutdown Whether to close the log file on shutdown
+   */
+  public void setCloseFileOnShutdown (boolean closeFileOnShutdown) {
+    this.closeFileOnShutdown = closeFileOnShutdown;
+  }
+
+  /**
+   * Sets the per-log-level hard timeouts for flushing.
+   * @param csvTimeouts A CSV string of kv-pairs. The key being the log-level in
+   * all caps and the value being the hard timeout for flushing in minutes. E.g.
+   * "INFO=30,WARN=10,ERROR=5"
+   */
+  public void setFlushHardTimeoutsInMinutes (String csvTimeouts) {
```

Comment:
Do we need to consider the invalid input as this method is public

Reply:
Ignores invalid input and printed out and error message.